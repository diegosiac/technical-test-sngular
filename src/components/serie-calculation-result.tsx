import { cn } from '@/utils/cn'

const LOADING_MESSAGE = 'Un momento, resolviendo cálculo...'

interface SerieCalculationResultProps {
  result: string | null
  loading: boolean
  errorMessage: string | null
}

export const SerieCalculationResult = ({
  result,
  loading,
  errorMessage,
}: SerieCalculationResultProps) => {
  const term = 'n'

  const showResultSection = result !== null || loading

  const resultContent = result ?? LOADING_MESSAGE

  return (
    <div className="flex flex-col gap-2" aria-live="polite">
      <label className="label">Operación:</label>
      <span className="text-xs bg-primary p-2 rounded-md">
        serie({term}) = triangular({term} * 2) - primo({term}) - fibonacci({term})
      </span>

      {showResultSection && (
        <>
          <label className="label mt-4">Resultado:</label>
          <span
            className={cn('bg-primary p-2 rounded-md text-xs break-words', {
              'animate-pulse': loading,
            })}
            aria-busy={loading}
          >
            {resultContent}
          </span>
        </>
      )}

      {errorMessage && (
        <span className="text-xs text-destructive" aria-live="polite">
          {errorMessage}
        </span>
      )}
    </div>
  )
}
