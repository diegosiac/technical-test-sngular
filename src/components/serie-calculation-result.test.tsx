import { render, screen } from '@testing-library/react'
import { describe, expect, it } from 'vitest'
import { SerieCalculationResult } from './serie-calculation-result'

describe('SerieCalculationResult Component', () => {
  it('should to match snapshot', () => {
    const { container } = render(
      <SerieCalculationResult result="200" loading={false} errorMessage={null} />
    )

    expect(container).toMatchSnapshot()
  })

  it('should not render the result label if there is no result', () => {
    render(<SerieCalculationResult result={null} loading={false} errorMessage={null} />)

    const resultLabel = screen.queryByLabelText('Resultado:')

    expect(resultLabel).not.toBeInTheDocument()
  })

  it('should render the result', () => {
    const expectedResult = '200'

    render(<SerieCalculationResult result={expectedResult} loading={false} errorMessage={null} />)

    const result = screen.getByText(expectedResult)
    expect(result).toBeInTheDocument()
  })

  it('should render the loading message', () => {
    render(<SerieCalculationResult result={null} loading errorMessage={null} />)

    const result = screen.getByText('Un momento, resolviendo cÃ¡lculo...')
    expect(result).toBeInTheDocument()
  })

  it('should render the error message', () => {
    const expectErrorMessage = 'Hubo un error al hacer el calculo'

    render(
      <SerieCalculationResult result={null} loading={false} errorMessage={expectErrorMessage} />
    )

    const result = screen.getByText(expectErrorMessage)

    expect(result).toBeInTheDocument()
  })
})
