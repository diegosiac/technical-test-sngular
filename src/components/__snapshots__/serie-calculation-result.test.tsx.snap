// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SerieCalculationResult Component > should to match snapshot 1`] = `
<div>
  <div
    aria-live="polite"
    class="flex flex-col gap-2"
  >
    <label
      class="label"
    >
      Operaci√≥n:
    </label>
    <span
      class="text-xs bg-primary p-2 rounded-md"
    >
      serie(
      n
      ) = triangular(
      n
       * 2) - primo(
      n
      ) - fibonacci(
      n
      )
    </span>
    <label
      class="label mt-4"
    >
      Resultado:
    </label>
    <span
      aria-busy="false"
      class="bg-primary p-2 rounded-md text-xs break-words"
    >
      200
    </span>
  </div>
</div>
`;
